File: 2025/01_secret_entrance/python/main.py
from pathlib import Path


def parse_doc(doc: str) -> list[int]:
    moves: list[int] = []
    for line in doc.splitlines():
        dir: int = -1 if line[0] == "L" else 1
        moves.append(dir * int(line[1:]))
    return moves


def count_zeros(moves: list[int]) -> int:
    pos = 50
    zero_count = 0
    for move in moves:
        pos += move
        pos %= 100
        if pos == 0:
            zero_count += 1
    return zero_count


def calculate_code(file: Path) -> int:
    return count_zeros(parse_doc(file.read_text()))


def count_zeros2(moves: list[int]) -> int:
    pos = 50
    zero_count = 0
    for move in moves:
        for _ in range(abs(move)):
            assert move != 0
            if move > 0:
                pos += 1
                if pos == 100:
                    pos = 0
            if move < 0:
                pos -= 1
                if pos == -1:
                    pos = 99
            if pos == 0:
                zero_count += 1
    return zero_count


def calculate_code2(file: Path) -> int:
    return count_zeros2(parse_doc(file.read_text()))


def main():
    print("first method: ", calculate_code(Path("../input1.txt")))
    print("second method: ", calculate_code2(Path("../input1.txt")))


if __name__ == "__main__":
    main()

File: 2025/01_secret_entrance/rust/src/main.rs
use rust::{calculate_code, calculate_code2};
use std::path::PathBuf;

fn main() {
    // The original script used a relative path `../input1.txt`.
    // We keep the same behaviour – the binary is expected to be run from
    // `mycode/target/debug` (or `release`), so `../input1.txt` points to the
    // file next to the Python version.
    let input_path = PathBuf::from("../input1.txt");

    println!("first method:  {}", calculate_code(&input_path));
    println!("second method: {}", calculate_code2(&input_path));
}
